### PR TITLE
[RFE] Drain nodes before deletion if requested

### PR DESCRIPTION
**Pull request by ingvagabund**
- drain nodes before deletion only if `drain-node: True` annotation is present on machine object)
- integrate the logic into the aws actuator Delete operation directly (until the node drain library is actually merged in k8s itself)

Continuing with work provided in https://github.com/openshift/cluster-api-provider-aws/pull/60

### CODE REVIEW

**Comment by ingvagabund** [comment]
```

File: Makefile [hunk 1]
```diff
@@ -102,7 +102,7 @@ k8s-e2e: ## Run k8s specific e2e test
 		-ginkgo.v
 
 .PHONY: test-e2e
-validate-e2e: ## Run e2e validation/gating test
+test-e2e: ## Run e2e validation/gating test
```
Comment (enxebre) [inline Makefile]: this does not belong to this commit. Is there a counterpart PR for prown?

**Comment by enxebre** [comment]
ping @wking

File: pkg/actuators/machine/actuator.go [hunk 2]
```diff
@@ -233,8 +240,55 @@ func (a *Actuator) Delete(context context.Context, cluster *clusterv1.Cluster, m
...
+}
+
+func (gl *glogLogger) Logf(format string, v ...interface{}) {
+	glog.Infof(format, v...)
+}
+
 // DeleteMachine deletes an AWS instance
 func (a *Actuator) DeleteMachine(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+	// Drain node before deleting
+	if machine.ObjectMeta.Annotations["drain-node"] == "True" && machine.Status.NodeRef != nil {
```
Comment (enxebre) [inline pkg/actuators/machine/actuator.go]: we should scope this similar to `cluster.k8s.io/machine`

File: pkg/actuators/machine/actuator.go [hunk 3]
```diff
@@ -233,8 +240,55 @@ func (a *Actuator) Delete(context context.Context, cluster *clusterv1.Cluster, m
...
 }
 
+type glogLogger struct{}
+
+func (gl *glogLogger) Log(v ...interface{}) {
+	glog.Info(v...)
+}
+
+func (gl *glogLogger) Logf(format string, v ...interface{}) {
+	glog.Infof(format, v...)
```
Comment (wking) [inline pkg/actuators/machine/actuator.go]: Doesn't save you much work, but you can import [`go-log/log/info`][1] instead of creating this local `glogLogger`.

[1]: https://godoc.org/github.com/go-log/log/info

File: Makefile (hunk 1 above)
Comment (ingvagabund) [reply Makefile]: ok, I will seperate it.

**Comment by wking** [comment]
Draining can also happen before a reboot, so it leaves a cordoned node.  If there something we should do to tell the control plane "we're deleting the node.  Go ahead and release its metadata, because it's not coming back"?  Not critical, but I don't want to leak cordoned nodes entries in scheduler memory (or wherever) as autoscalers shuffle node allocations.

File: pkg/actuators/machine/actuator.go (hunk 2 above)
Comment (ingvagabund) [reply pkg/actuators/machine/actuator.go]: As long as the `drain-node` label is not upstream accepted, I would not issue `cluster.k8s.io` prefix. Plus, the label will get removed anyway. Though, we can do `openshift.io/drain-node` for now. Hmm?

File: pkg/actuators/machine/actuator.go (hunk 3 above)
Comment (ingvagabund) [reply pkg/actuators/machine/actuator.go]: The goal was to use only a single logger wherever possible. I did not want to introduce another one. Thus, creating a special (not super cool at all) data type just for this.

**Comment by ingvagabund** [comment]
> Draining can also happen before a reboot, so it leaves a cordoned node. If there something we should do to tell the control plane "we're deleting the node. Go ahead and release its metadata, because it's not coming back"? Not critical, but I don't want to leak cordoned nodes entries in scheduler memory (or wherever) as autoscalers shuffle node allocations.

The control plane can listen for "Deleted Node ??? drained" events. Would that be enough?

File: pkg/actuators/machine/actuator.go (hunk 3 above)
Comment (wking) [reply pkg/actuators/machine/actuator.go]: > The goal was to use only a single logger...

So you're avoiding `info.New(glog.V(0))`?

[1]: https://github.com/go-log/log/blob/master/info/info.go#L24

**Comment by wking** [comment]
> The control plane can listen for "Deleted Node ??? drained" events. Would that be enough?

Maybe?  I dunno what the usual approach is, I'm just hoping there is one ;).

File: pkg/actuators/machine/actuator.go (hunk 3 above)
Comment (ingvagabund) [reply pkg/actuators/machine/actuator.go]: Did not know you can do it this way. On it.

**Comment by ingvagabund** [comment]
> Maybe? I dunno what the usual approach is, I'm just hoping there is one ;).

The node draining is disabled by default. So this can be good opportunity to see what happens when a node is drained and restarted before the draining is finished.

**Comment by wking** [comment]
> So this can be good opportunity to see what happens when a node is drained and restarted before the draining is finished.

I'm not worried about strandings pods (I expect the scheduler to be rubust on that front).  I'm concerned about storing "there's a cordoned node at `$IP`" information forever, even when *we* know that node is never coming back.

**Comment by ingvagabund** [comment]
> I'm not worried about strandings pods (I expect the scheduler to be rubust on that front). I'm concerned about storing "there's a cordoned node at $IP" information forever, even when we know that node is never coming back.

imho, that's up to the node controller to react properly, the node object will disappear after some time which will free the object.

File: Makefile (hunk 1 above)
Comment (ingvagabund) [reply Makefile]: Done through https://github.com/openshift/cluster-api-provider-aws/pull/136

File: pkg/actuators/machine/actuator.go [hunk 4]
```diff
@@ -233,8 +240,55 @@ func (a *Actuator) Delete(context context.Context, cluster *clusterv1.Cluster, m
...
+				glog.Warningf("drain failed for machine %s: %v", machine.Name, err)
+			} else {
+				glog.Infof("drain successful for machine %s", machine.Name)
+				a.eventRecorder.Eventf(machine, corev1.EventTypeNormal, "Deleted", "Node %q drained", node.Name)
+			}
+		} else {
+			glog.Infof("Missing client config, unable to build kube client")
+		}
+	} else {
+		glog.Infof("found force-deletion label, skipping node drain")
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/actuators/machine/actuator.go]: Removed completely

File: test/e2e/provider_expectations.go [hunk 5]
```diff
@@ -198,3 +199,119 @@ MachineLoop:
...
+		if machineDeletedEvent == nil {
+			glog.Infof("Unable to find %q machine deleted event", machine.Name)
+			return false, nil
+		}
+
+		glog.Infof("Node %q drained event recorded: %#v", nodeName, *nodeDrainedEvent)
+
+		if machineDeletedEvent.FirstTimestamp.Before(&nodeDrainedEvent.FirstTimestamp) {
+			err := fmt.Errorf("Machine %q deleted before node %q got drained", machine.Name, nodeName)
+			glog.Info(err)
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: why not just `glog.errorf`?

File: test/e2e/provider_expectations.go [hunk 6]
```diff
@@ -198,3 +199,119 @@ MachineLoop:
...
+					return true, nil
+				}
+			}
+		}
+		return false, fmt.Errorf("No machine found with NodeRef not set")
+	}); err != nil {
+		return err
+	}
+
+	glog.Info("Get nodeList")
```
Comment (enxebre) [inline test/e2e/provider_expectations.go]: why do you need to get the node? if there's actually a reason you couldn't you just fetch it right away instead of get the list?

File: test/e2e/provider_expectations.go (hunk 5 above)
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: Ha, yes, it's suppose to be `glog.Errorf`

File: test/e2e/provider_expectations.go [hunk 7]
```diff
@@ -198,3 +199,119 @@ MachineLoop:
...
+	machine.ObjectMeta.Annotations["openshift.io/drain-node"] = "True"
+	if err := tc.client.Update(context.TODO(), &machine); err != nil {
+		return fmt.Errorf("Unable to set `drain-node` annotation")
+	}
+
+	currentTimestamp := metav1.Now()
+
+	glog.Info("Delete machine and observe node draining")
+	if err := tc.client.Delete(context.TODO(), &machine); err != nil {
+		return fmt.Errorf("Unable to delete %q machine", machine.Name)
```
Comment (frobware) [inline test/e2e/provider_expectations.go]: nit: "unable to delete machine %q"

File: pkg/actuators/machine/actuator.go [hunk 8]
```diff
@@ -233,8 +241,54 @@ func (a *Actuator) Delete(context context.Context, cluster *clusterv1.Cluster, m
...
+			&kubedrain.DrainOptions{
+				Force:              true,
+				IgnoreDaemonsets:   true,
+				DeleteLocalData:    true,
+				GracePeriodSeconds: -1,
+				Logger:             info.New(glog.V(0)),
+			},
+		); err != nil {
+			// Machine still tries to terminate after drain failure
+			glog.Warningf("drain failed for machine %s: %v", machine.Name, err)
```
Comment (frobware) [inline pkg/actuators/machine/actuator.go]: nit: everywhere else (I think?) you use `%q` for names.

File: test/e2e/provider_expectations.go (hunk 6 above)
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: Just following the same pattern as is in the previous tests.

File: pkg/actuators/machine/actuator.go [hunk 9]
```diff
@@ -233,8 +241,54 @@ func (a *Actuator) Delete(context context.Context, cluster *clusterv1.Cluster, m
...
+			glog.Error(err)
+			return err
+		}
+		kubeClient, err := kubernetes.NewForConfig(a.config)
+		if err != nil {
+			return fmt.Errorf("unable to build kube client: %v", err)
+		}
+		node, err := kubeClient.CoreV1().Nodes().Get(machine.Status.NodeRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("unable to get %q node: %v", machine.Status.NodeRef.Name, err)
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply pkg/actuators/machine/actuator.go]: This block executes only when `machine.Status.NodeRef != nil`. See https://github.com/openshift/cluster-api-provider-aws/pull/135/commits/656e221c53ff754ad7164b9124dd237e885126dd#diff-acdcd15b24a848e93a682408b101755fR257

File: test/e2e/provider_expectations.go [hunk 10]
```diff
@@ -198,3 +199,119 @@ MachineLoop:
...
+func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
+	listOptions := client.ListOptions{
+		Namespace: namespace,
+	}
+
+	var machine capiv1alpha1.Machine
+	var nodeName string
+	var node *corev1.Node
+
+	glog.Info("Get machineList with at least one machine with NodeRef set")
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: The draining is performed only when `machineItem.Status.NodeRef` is not nil.

File: test/e2e/provider_expectations.go [hunk 11]
```diff
@@ -198,3 +199,119 @@ MachineLoop:
...
+	if err := wait.PollImmediate(1*time.Second, waitShort, func() (bool, error) {
+		machineList := capiv1alpha1.MachineList{}
+		if err := tc.client.List(context.TODO(), &listOptions, &machineList); err != nil {
+			glog.Errorf("error querying api for machineList object: %v, retrying...", err)
+			return false, nil
+		}
+		for _, machineItem := range machineList.Items {
+			// empty or non-worker role skipped
+			if machineItem.Labels["sigs.k8s.io/cluster-api-machine-role"] == "worker" {
```
[the remark this answers is not in the document]
Comment (ingvagabund) [reply test/e2e/provider_expectations.go]: In general `!master` != `worker`. Right now it holds, though in future we might have `compute` as well. So targeting `worker` seems enough for now.

**Comment by ingvagabund** [comment]
Thanks guys, I will squash and rebase. Then relabel.

File: Gopkg.lock [hunk 12]
```diff
@@ -99,6 +99,17 @@
   revision = "0ca9ea5df5451ffdf184b4428c902747c2c11cd7"
   version = "v1.0.0"
```
Comment (enxebre) [inline Gopkg.lock]: Please as a general good practice let's include reason for re vendoring in the commit message

Comment (ingvagabund) [reply Gopkg.lock]: `Gopkg.lock` is auto-generated by godep

**Comment by ingvagabund** [comment]
Just removing the time constraint of event. It's likely the node drained event, resp. node delete event can have earlier timestamp than the one captured in the tests. E.g. machine controller container time and test container time does not have to be in sync.
